packages/core/src/behaviors.ts

```diff
@@ -0,0 +1,123 @@
+import { ActorContext, ActorRef, Behavior, EventObject, Observer } from '.';
+import { doneInvoke, error } from './actions';
+import { toActorRef } from './Actor';
+import { toObserver } from './utils';
+
+/**
+ * Returns an actor behavior from a reducer and its initial state.
+ *
+ * @param transition The pure reducer that returns the next state given the current state and event.
+ * @param initialState The initial state of the reducer.
+ * @returns An actor behavior
+ */
+export function fromReducer<TState, TEvent extends EventObject>(
+  transition: (
+    state: TState,
+    event: TEvent,
+    actorContext: ActorContext<TEvent, TState>
+  ) => TState,
+  initialState: TState
+): Behavior<TEvent, TState> {
+  return {
+    transition,
+    initialState
+  };
+}
+
+type PromiseEvents<T> =
+  | { type: 'fulfill'; data: T }
+  | { type: 'reject'; error: unknown };
+
+export function fromPromise<T>(
+  promiseFn: () => Promise<T>
+): Behavior<PromiseEvents<T>, T | undefined> {
+  return {
+    transition: (state, event, { parent, id, observers }) => {
+      switch (event.type) {
+        case 'fulfill':
+          parent?.send(doneInvoke(id, event.data));
+          return event.data;
+        case 'reject':
+          parent?.send(error(id, event.error));
+          observers.forEach((observer) => {
+            observer.error(event.error);
+          });
+          return undefined;
```

Comment:
isn't returning `undefined` here meaning that we won't be able to display the error in a component when using smth like this?
```js
const actorRef = useSpawn(fromPromise(() => Promise.reject(new Error('Foo.'))));
const [state] = useActor(actorRef);
```

should we allow it though? if not - how the user should handle those errors in the component's context

Reply:
By using an observer:

```js
useEffect(() => {
  actorRef.subscribe({ error: e => setError(e) });
}, [actorRef]);
```

Reply:
isn't this boilerplate-y though? as a user I would expect error handling to be more of primary concern and not something that I have to wire up myself. By making it harder to handle this we risk people not handling errors as people are lazy (me included 😉 ). Some will handle them, some will not - the API doesn't make it easy though, it doesn't create a pit of success here.

Reply:
What would a more convenient API look like? What we _can't_ do is shove the error in the `state`; the `state` is meant to represent the emitted value.

Maybe we can do something like this:
```ts
const [state, send, { error }] = useActor(actorRef);
```

Reply:
Another alternative:
```js
const error = useActorError(actorRef); // usually `undefined`
```

Reply:
We might want to throw an error during rendering - which would require wrapping the internal state of `useActor` into a wrapper object, subscribing to errors like you have showcased here and throwing during rerender. Similarly to how, for example, `react-query` handles that:
https://github.com/tannerlinsley/react-query/blob/16b7d290c70639b627d9ada32951d211eac3adc3/src/react/useBaseQuery.ts#L124-L131

Reply:
@Andarist I'm revisiting this, because you're right, the way we're representing promise state is a bit awkward.

What are your thoughts on this?

```js
const [{ data, error, status }, send] = useActor(fromPromise(/* ... */));
```

Reply:
Made the changes here: 30c310c

Reply:
Oh, I have sort of commented on that a minute ago on Discord so let me just copy-paste my comment from there:
> Huh, a lot of catching up here - I was thinking about the error stuff as well, just didnt have time to write down my thoughts here. I came to the conclusion that, in my opinion, we should throw errors and let error boundaries catch them - what @mattpocock#9242 said initially. This solution~ might be framework-specific, for now I'm thinking mostly about React and `@xstate/react`. I think it's reasonable for APIs of framework integration packages to diverge because they should feel native to the users of those frameworks, exact API parity between them is nice but imho a conceptual non-goal.

> I'm not saying that API definitely has to diverge, this would have to be explored - individually for each framework. I just lack experience with Vue etc so I can't say with any certainty how what I'm saying applies to those frameworks.

> It might feel a little bit unnatural but I think it's the reasonable choice for React. Note that there is never anything stopping people from normalizing "errors" into states~ and just having "error data" available as the first item returned from `useActor`. 

> A big problem with returning rejection/error values anyhow from any hook is that... errors are basically untyped in TS, so this doesn't provide any actual type safety and is unsound by definition. Guiding people towards normalizing their errors into "states" is creating a pit of success - they have to type this and normalize their error data into that expected shape. It's clear where and how one should do it - just add `.catch(() => ...)` and do it there, whatever you return from there has to match the expected type. It IMHO feels nice and is what I would, as a user, prefer.

> And unhandled errors would just go to an error boundary where it's expected that the received error value is any/unknown and that it has to be refined there.

> This is also similar to what React Query does - however, it's currently opt-in and called as experimental: https://react-query.tanstack.com/guides/suspense